format.go

```diff
@@ -350,3 +350,80 @@ func Glob(input, match string) bool {
 	// Check suffix last.
 	return trailingGlob || strings.HasSuffix(input, parts[last])
 }
+
+const maxWordSplitLength = 30
+
+// splitAtWord is a text splitter that takes into consideration a few things:
+//    * Ensuring the returned text is no longer than maxWidth.
+//    * Attempting to split at the closest word boundary, while still staying inside
+//      of the specific maxWidth.
+//    * if there is no good word boundry for longer words (or e.g. links, raw data, etc)
+//      that are above maxWordSplitLength characters, split the word into chunks to fit the
+// maximum width.
+func splitAtWord(input string, maxWidth int) (output []string) {
+	// TODO: breaks multi-spaces.
+	words := strings.Fields(input)
+
+	// TODO: don't split a url if there isn't enough space, if it can safely fit within
+	// the next line.
+	// TODO: also split on newline, if splitting is enabled? makes it easier to just
+	// pipe text in.
+	// TODO: if word contains a dash, and adding the word to the line causes an overflow,
+	// try to split on the dash?
+
+	// Increment maxWidth for calculations, because we always prefix with a space (then
+	// strip it before we return).
+	maxWidth++
+
+	var i, spaceRemaining int
+	var split string
+
+setupNextSplit:
+	spaceRemaining = maxWidth
+	split = ""
+beginLoop:
+	for i < len(words) {
+		// Last line was the perfect length, add to output and keep looping.
+		if spaceRemaining == 0 {
+			output = append(output, split[1:])
+			goto setupNextSplit
+		}
+
+		// Word makes the line too long.
+		if len(words[i])+1 > spaceRemaining {
+			// Is the word small enough to where we don't need to split it up?
+			if len(words[i]) < maxWordSplitLength && maxWidth >= maxWordSplitLength {
+				output = append(output, split[1:])
+				goto setupNextSplit
+			}
+
+			split += " " + words[i][0:spaceRemaining-1]
+			if len(words) == i {
+				words = append(words, words[i][spaceRemaining-1:])
+			} else {
+				words = append(words[:i+1], words[i:]...)
+				words[i+1] = words[i][spaceRemaining-1:]
+				words[i] = words[i][0 : spaceRemaining-1]
+			}
+			spaceRemaining -= 1 + len(words[i][0:spaceRemaining-1])
```

Comment:
This does not seem to handle multibyte characters correctly. Consider the following test case from my proposal:

https://github.com/lrstanley/girc/blob/0503d81bfbf5740eedd6b3c4cb2b94e3ee6382b1/split_test.go#L23

If I am not mistaken this should translate to your implementation as:

```go
result := splitAtWord("ξ_λχ", len("λχ"))
```

For the given invocation, your implementation returns `[,ξ_", "χ"]` as the result, discarding the lambda (`λ`) character which seems wrong to me unless I am missing something here.

Reply:
I am planning on figuring this out when writing the tests.